source/game/system/GhostFile.cc

```diff
@@ -0,0 +1,166 @@
+    m_driftIsAuto = raw->parseAt<u8>(0xD) >> 1 & 0x1;
+}
+
+Character GhostFile::character() const {
+    return m_character;
+}
+
+Vehicle GhostFile::vehicle() const {
+    return m_vehicle;
+}
+
+Course GhostFile::course() const {
+    return m_course;
+}
+
+const u8 *GhostFile::inputs() const {
+    return m_inputs;
+}
+
+bool GhostFile::driftIsAuto() const {
+    return m_driftIsAuto;
+}
+
+Timer::Timer() : min(0), sec(0), mil(0) {}
+
+Timer::Timer(u32 data) {
+    min = static_cast<u16>(data >> 0x19);
+    sec = static_cast<u8>(data >> 0x12) & 0x7F;
+    mil = static_cast<u16>(data >> 8) & 0x3FF;
+}
+
+Timer::~Timer() = default;
+
+RawGhostFile::RawGhostFile(const u8 *rkg) {
+    if (!isValid(rkg)) {
+        K_PANIC("Invalid RKG header");
+    }
+
+    if (compressed(rkg)) {
+        if (!decompress(rkg)) {
+            K_PANIC("Failed to decompress RKG!");
+        }
+    } else {
+        memcpy(m_buffer, rkg, 0x27FC);
+    }
+}
+
+RawGhostFile::~RawGhostFile() = default;
+
+bool RawGhostFile::decompress(const u8 *rkg) {
+    // Copy header
+    memcpy(m_buffer, rkg, 0x88);
+
+    // Unset compressed flag
+    *(m_buffer + 0xC) &= 0xF7;
+
+    // Get uncompressed size
+    s32 uncompressedSize = EGG::Decomp::GetExpandSize(rkg + 0x8C);
+
+    if (uncompressedSize == 0 || uncompressedSize > 0x2774) {
+        return false;
+    }
+
+    EGG::Decomp::DecodeSZS(rkg + 0x8C, m_buffer + 0x88);
+
+    // Set input data section length
+    *reinterpret_cast<u16 *>(m_buffer + 0xE) = static_cast<u16>(uncompressedSize);
+
+    return true;
+}
+
+bool RawGhostFile::isValid(const u8 *rkg) const {
+    if (strncmp(reinterpret_cast<const char *>(rkg), "RKGD", 4) != 0) {
+        K_PANIC("RKG header malformed");
+        return false;
+    }
+
+    u32 ids = parse<u32>(*reinterpret_cast<const u32 *>(rkg + 0x8));
+    Vehicle vehicle = static_cast<Vehicle>(ids >> 0x1a);
+    Character character = static_cast<Character>((ids >> 0x14) & 0x3f);
+    u8 year = (ids >> 0xd) & 0x7f;
+    u8 day = (ids >> 0x4) & 0x1f;
+    u8 month = (ids >> 0x9) & 0xf;
+
+    if (vehicle >= Vehicle::Max || character >= Character::Max) {
+        return false;
+    }
+
+    if (year >= 100 || day >= 32 || month > 12) {
+        return false;
+    }
+
+    // Validate weight class match
+    WeightClass charWeight = characterToWeight(character);
+    WeightClass vehicleWeight = vehicleToWeight(vehicle);
+    if (charWeight != vehicleWeight) {
+        K_PANIC("Character/Bike weight class mismatch!");
+    }
+
+    // TODO:
+    // - Check for valid controller type??
+    // - Check lap times sum to race time??
+
+    return true;
+}
+
+const u8 *RawGhostFile::buffer() const {
+    return m_buffer;
+}
+
+template <typename T>
+T RawGhostFile::parseAt(size_t offset) const {
+    return parse<T>(*reinterpret_cast<const T *>(m_buffer + offset));
+}
+
+bool RawGhostFile::compressed(const u8 *rkg) const {
+    return ((*(rkg + 0xC) >> 3) & 1) == 1;
```

Comment:
Are you sure we compare against 1, and it isn't a comparison against 0 in the base game?

Reply:
In `RawGhostFile::compress` the check is performed via `rlwinm r0, r0, 0x15, 0x1f, 0x1f` and `cmplwi r0, 0x1`. I take this to mean that it's a comparison against 1.

Reply:
You assume correctly. I would keep the explicit check against 1 then.